runtime-modules/bounty/src/lib.rs

```diff
@@ -930,33 +1100,45 @@ decl_module! {
             let bounty = Self::ensure_bounty_exists(&bounty_id)?;
 
             let current_bounty_stage = Self::get_bounty_stage(&bounty);
-            Self::ensure_bounty_stage(current_bounty_stage, BountyStage::FailedBountyWithdrawal)?;
 
-            ensure!(
-                <BountyContributions<T>>::contains_key(&bounty_id, &funder),
-                Error::<T>::NoBountyContributionFound,
-            );
+            let funding = Self::ensure_bounty_contribution_exists(&bounty_id, &funder)?;
 
-            let funding_amount = <BountyContributions<T>>::get(&bounty_id, &funder);
-            let cherry_fraction = Self::get_cherry_fraction_for_member(&bounty, funding_amount);
-            let withdrawal_amount = funding_amount + cherry_fraction;
+            let bounty_creator_manager = BountyActorManager::<T>::get_bounty_actor_manager(
+                bounty.creation_params.creator.clone(),
+            )?;
 
             //
             // == MUTATION SAFE ==
             //
 
-            bounty_funder_manager.transfer_funds_from_bounty_account(bounty_id, withdrawal_amount)?;
-
-            <BountyContributions<T>>::remove(&bounty_id, &funder);
-
-            Self::deposit_event(RawEvent::BountyFundingWithdrawal(bounty_id, funder));
+            match current_bounty_stage{
+                BountyStage::FailedBountyWithdrawal => {
+                    Self::withdraw_funding_mutation(
+                        &bounty_id,
+                        &bounty,
+                        funder,
+                        &bounty_funder_manager,
+                        funding);
+                },
+                BountyStage::SuccessfulBountyWithdrawal => {
+                    Self::withdraw_funding_state_bloat_bond_mutation(
+                        &bounty_id,
+                        funder,
+                        &bounty_funder_manager,
+                        funding);
+                },
+                _ => fail!(Self::unexpected_bounty_stage_error(current_bounty_stage))
```

Comment:
If the stage is not one of the above, the function fails safely without any mutation,
There are invalid stages that can't reach `unexpected_bounty_stage_error`, for example, FundingExpired and Funding without contributions, since there are no contributions, this condition 
`let funding = Self::ensure_bounty_contribution_exists(&bounty_id, &funder)?;` will fail first.